### PR TITLE
networkd: add route expiration handler

### PR DESCRIPTION
**Pull request by ssahani**
Fix for #3232

### CODE REVIEW

File: src/network/networkd-route.c [hunk 1]
```diff
@@ -407,21 +407,53 @@ int route_remove(Route *route, Link *link,
         if (r < 0)
                 return log_error_errno(r, "Could not send rtnetlink message: %m");
 
+        /* route may not be exist in kernel. If we fail still remove it */
+        route_free(route);
```
Comment (keszybz) [inline src/network/networkd-route.c]: It looks really strange for route_remove to free route.

File: src/network/networkd-route.c [hunk 2]
```diff
@@ -407,21 +407,53 @@ int route_remove(Route *route, Link *link,
...
         int r;
 
         assert(route);
 
-        r = route_remove(route, route->link, NULL);
+        r = route_remove(route, route->link, route_expire_callback);
         if (r < 0)
                 log_warning_errno(r, "Could not remove route: %m");
 
+        route->link->link_messages++;
```
Comment (keszybz) [inline src/network/networkd-route.c]: ... and here you use `route` after freeing it.

File: src/network/networkd-route.c [hunk 3]
```diff
@@ -407,21 +407,53 @@ int route_remove(Route *route, Link *link,
...
+
+        link_ref(route->link);
```
Comment (keszybz) [inline src/network/networkd-route.c]: Can you explain why link_ref is called and where the matching link_unref happens?

File: src/network/networkd-route.c (hunk 2 above)
Comment (ssahani) [inline src/network/networkd-route.c]: yikes fixed it .

**Pull request merged**